File: leetcode-solutions/0124-binary-tree-maximum-path-sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxi=INT_MIN;
    int maxdepth(TreeNode* root){
        if(root==nullptr)return 0;
        int lh=max(0,maxdepth(root->left));
        int rh=max(0,maxdepth(root->right));
        maxi=max(maxi,root->val+lh+rh);
        return root->val+max(lh,rh);
    }
    int maxPathSum(TreeNode* root) {
        maxdepth(root);
        return maxi;
    }
};

File: leetcode-solutions/0124-binary-tree-maximum-path-sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxi=Integer.MIN_VALUE;
    public int height(TreeNode root){
        if(root==null)return 0;
        int left=height(root.left);
        if(left<0)left=0;
        int right=height(root.right);
        if(right<0)right=0;
        maxi=Math.max(root.val+left+right,maxi);
        return root.val+Math.max(left,right);
    }
    public int maxPathSum(TreeNode root) {
        height(root);
        return maxi;
    }
}
